core-rust/state-manager/src/metrics.rs

```diff
@@ -394,3 +414,64 @@ impl MetricLabel for VertexPrepareStopReason {
         }
     }
 }
+
+/// A number of most recent [`RoundSlot`]s of a single validator to track for metrics purposes.
+const PROPOSAL_HISTORY_LEN: usize = 100;
+
+/// An indication of (any kind of) missed round vs successful round of a validator.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+enum RoundSlot {
+    Success,
+    Missed,
+}
+
+/// A higher-level metric helper, tracking a number of recent proposal misses of a specific
+/// validator.
+pub struct SelfProposalMissTracker {
+    buffer: Mutex<RingBuffer<RoundSlot, PROPOSAL_HISTORY_LEN>>,
```

Comment:
Why is `buffer` behind a mutex?

Reply:
as discussed on Slack:

We want interior mutability here (as is the case with all other metric bumps), and we don't feel too comfortable with `unsafe`, so we prefer a hopefully-cheap mutex (which will never block in practice).